File: algorithms_c/Valve Insight PlatformIO/lib/switchingDetection/src/switchingDetectionOnline.c
#include "switchingDetectionOnline.h"


#include <math.h>
#include <string.h>

#define TAIL_WINDOW_SIZE          200

static size_t n;
static float integral = 0;
static int32_t valuesSum = 0;
static float tailWindow[TAIL_WINDOW_SIZE];


float round2(float x) {
    return round(x * 100.0) / 100.0;
}

void switchingDetectionOnline_Reset(void) {
	n = 0;
	integral = 0;
	valuesSum = 0;
	memset(tailWindow, 0, sizeof(tailWindow));
}

float switchingDetectionOnline_Calculate(int16_t adcValue, float timestep) {
	// running integral
	valuesSum += adcValue;
	
	tailWindow[n % TAIL_WINDOW_SIZE] = adcValue;
	++n;

	// Calculate tail mean
	float tailMean = 0;
	if (n > TAIL_WINDOW_SIZE) {
		uint16_t i;
		for (i=0; i < TAIL_WINDOW_SIZE; ++i) {
			tailMean += tailWindow[i];
		}
		tailMean /= TAIL_WINDOW_SIZE;
	}else{
		// Not a full tail window acquired yet. Stop here.
		return 0;
	}

	float maxIntegral = n * timestep;
	float currIntegral = maxIntegral - maxIntegral - (float)valuesSum * timestep / tailMean;
	currIntegral = fabs(currIntegral);
	currIntegral = round2(currIntegral);

	return currIntegral;
}

File: algorithms_python/src/valveinsight/switching_detection_online.py
import numpy as np

class switching_detection_online:
    """
    Online estimation of the integral value for successful switching. 
    If the tail_window if large enough to cover the expected switching duration, 
    you can stop the data acquisition after the integral value is large enough. 
    """
    
    def __init__(self):
        self.integral = 0
        self.n = 0
        self.tail_window = np.zeros(200)

    def switching_detection_online(self, current_mA:float, timestep_ms:float):
        self.integral += current_mA * timestep_ms
        self.tail_window[self.n % len(self.tail_window)] = current_mA
        self.n += 1        

        # Calculate tail mean
        if self.n > len(self.tail_window):
            tail_mean = np.mean(self.tail_window)
        else:
            # Not a full tail window acquired yet
            return 0, 0

        max_integral = self.n * timestep_ms
        current_integral = max_integral - self.integral / tail_mean
        current_integral = np.round(current_integral, 2)
        return current_integral, tail_mean